snowfall/training/ctc_graph.py

```diff
@@ -0,0 +1,159 @@
+# Copyright (c)  2020  Xiaomi Corp.       (author: Fangjun Kuang)
+
+from functools import lru_cache
+from typing import (
+    Iterable,
+    List,
+)
+
+import torch
+import k2
+import _k2
+
+
+def build_ctc_topo(phones: List[int]) -> k2.Fsa:
+    '''Build CTC topology.
+
+    The resulting topology converts repeated input
+    symbols to a single output symbol.
+
+    Args:
+      phones:
+        A list of phones
+    Returns:
+      Returns an FSA that converts repeated symbols to a single symbol.
+    '''
+    num_states = len(phones)
+    final_state = num_states
+    rules = ''
+    for i in range(num_states):
+        for j in range(num_states):
+            if i == j:
+                rules += f'{i} {i} {phones[i]} 0 0.0\n'
+            else:
+                rules += f'{i} {j} {phones[j]} {phones[j]} 0.0\n'
+        rules += f'{i} {final_state} -1 -1 0.0\n'
+    rules += f'{final_state}'
+    ans = k2.Fsa.from_str(rules)
+    return ans
+
+
+def compose(a_fsa: k2.Fsa, b_fsa: k2.Fsa) -> k2.Fsa:
+    '''Compose two graphs.
+
+    Args:
+      a_fsa:
+        The FSA on the left hand side.
+      b_fsa:
+        The FSA on the right hand side.
+
+    Returns:
+      The composition result. It will have:
```

Comment:
`It will have:` what? seems  there were some documentation?

Reply:
Fixed.